### PR TITLE
feat: add method to get all events emitted by the contract

### LINKED ISSUE #55: [Feature]: add function that returns a list of all events emitted by a contract
### What is the feature you would like to see?

From https://github.com/OpenZeppelin/stylus-test-helpers/pull/34#discussion_r1967227435

### Contribution Guidelines

- [x] I agree to follow this project's Contribution Guidelines

### PR DESCRIPTION
**Pull request by isSerge**
Resolves #55 

#### PR Checklist

- [x] Tests
- [x] Documentation
- [x] Changelog

### CODE REVIEW

**Comment by 0xNeshi** [comment]
Hi @isSerge and thanks for the contribution. We are currently under a huge workload, so we will be reviewing the PR as soon as we can.

File: crates/motsu/src/context.rs [hunk 1]
```diff
@@ -905,6 +905,21 @@ impl<ST: StorageType + Router + 'static> Contract<ST> {
...
+    #[must_use]
+    pub fn all_events(&self) -> Vec<LogData> {
+        let context = VM::context();
+        context
+            .storage()
+            .persistent
+            .contracts
+            .get(&self.address)
+            .map(|contract_storage| contract_storage.events.clone())
+            .unwrap_or_default()
```
Comment (0xNeshi) [inline crates/motsu/src/context.rs:921]: This should be a `pub(crate)` function inside `VM` and called like:
```rust
   #[must_use]
    pub fn all_events(&self) -> Vec<LogData> {
        VM::context().all_events_for(&self.address)
    }
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -29,6 +29,7 @@ The format is based on [Keep a Changelog], and this project adheres to [Semantic
 - Support `fallback` and `receive` functions. #88
 - Derived `PartialEq`, `Eq` and `Debug` for `Account`. #81
 - Enabled additional conversions between `PrivateKeySigner`, `Address` and `Account`. #84
+- Added `Contract::all_events` method to retrieve all events emitted by the contract #95
```
Comment (0xNeshi) [inline CHANGELOG.md:32]:
```suggestion
- Added `Contract::all_events` method to retrieve all events emitted by the contract. #95
```

File: crates/motsu/src/lib.rs [hunk 3]
```diff
@@ -379,6 +379,110 @@ mod ping_pong_tests {
...
+
+        // make successful call
+        ping.sender(alice).ping(pong.address(), TEN).motsu_unwrap();
+
+        // Check both events
+        let ping_events = ping.all_events();
+        let pong_events = pong.all_events();
+
+        assert_eq!(ping_events.len(), 1);
+        assert_eq!(pong_events.len(), 1);
```
Comment (0xNeshi) [inline crates/motsu/src/lib.rs:484]: This is actually a very good test case.

Could you add after this successful call an additional call that will revert, to ensure that once any event is added, the reverts still do not add new events? You can copy/paste the previous revert call.

File: crates/motsu/src/lib.rs [hunk 4]
```diff
@@ -379,6 +379,110 @@ mod ping_pong_tests {
...
+
+        let pong_event_2 =
+            Ponged::decode_log_data(&pong_events[1], true).unwrap();
+
+        assert_eq!(pong_event_2.from, ping.address());
+        assert_eq!(pong_event_2.value, TEN + ONE);
+    }
+
+    #[motsu::test]
+    fn assert_revert_all_event(
```
Comment (0xNeshi) [inline crates/motsu/src/lib.rs:439]: Add additional tests covering edge cases:
- multiple contracts of the same type but different addresses (e.g. two `Ping` contracts) should store their events independently.
- emit multiple events on the same contract and ensure all of them are stored in the order they were emitted.
- multiple calls to `all_events` should return the exact same events in the exact same order.
- update test [storage_cleanup](https://github.com/OpenZeppelin/stylus-test-helpers/blob/978e8a6ffdfe0d8f505d99dc07d6fbb68b4a2009/crates/motsu/src/lib.rs#L304) to include a call to `all_events`, ensuring the recreated contract returns no events.

File: crates/motsu/src/lib.rs [hunk 5]
```diff
@@ -379,6 +387,346 @@ mod ping_pong_tests {
...
+    // multiple instances of the same contract type
+    #[motsu::test]
+    fn all_events_multiple_instances_same_type(
+        pong: Contract<PongContract>,
+        alice: Address,
+        bob: Address,
+    ) {
+        // Create two PingContract instances at different addresses
+        let ping1 = Contract::<PingContract>::new();
+        let ping2 = Contract::<PingContract>::new();
```
Comment (0xNeshi) [inline crates/motsu/src/lib.rs:518]:
```suggestion
    fn all_events_multiple_instances_same_type(
        ping1: Contract<PingContract>,
        ping2: Contract<PingContract>,
        pong: Contract<PongContract>,
        alice: Address,
        bob: Address,
    ) {
```
Couldn't we initiate pings like this?

File: crates/motsu/src/lib.rs [hunk 6]
```diff
@@ -379,6 +387,346 @@ mod ping_pong_tests {
...
+
+        // Make sure addresses are different
+        assert_ne!(
+            ping1.address(),
+            ping2.address(),
+            "Ping contracts should have different addresses"
+        );
```
Comment (0xNeshi) [inline crates/motsu/src/lib.rs:525]: They will have different addresses by definition, this check can be removed

**Review by 0xNeshi — APPROVED** [review]
Excellent work @isSerge :fire:

Left some nits, we can merge after fixing them

File: crates/motsu/src/context.rs [hunk 7]
```diff
@@ -552,6 +553,18 @@ impl VM {
         let mut storage = self.storage();
         storage.chain_id = chain_id;
     }
+
+    /// Get all events emitted by the contract at `address`.
+    /// Returns a vector of `LogData` objects representing the events emitted by
+    /// the contract.
+    pub(crate) fn all_events_for(&self, address: &Address) -> Vec<LogData> {
```
Comment (bidzyyys) [inline crates/motsu/src/context.rs:560]: Clippy warning abou `&self`

**Review by bidzyyys — COMMENTED** [review]
Please apply clippy warnings and @0xNeshi comments.

File: crates/motsu/src/lib.rs [hunk 8]
```diff
@@ -379,6 +387,337 @@ mod ping_pong_tests {
...
+        assert!(
+            pong_events_after_revert.is_empty(),
+            "Pong should have no events after revert"
+        );
+
+        // make successful call
+        ping.sender(alice).ping(pong.address(), TEN).motsu_unwrap();
+
+        // Check both events
```
Comment (bidzyyys) [inline crates/motsu/src/lib.rs]:
```suggestion
        // Check both events.
```

File: crates/motsu/src/lib.rs [hunk 9]
```diff
@@ -379,6 +387,337 @@ mod ping_pong_tests {
...
+    // multiple instances of the same contract type
+    #[motsu::test]
+    fn all_events_multiple_instances_same_type(
+        ping1: Contract<PingContract>,
+        ping2: Contract<PingContract>,
+        pong: Contract<PongContract>,
+        alice: Address,
+        bob: Address,
+    ) {
+        // Initial state: no events
```
Comment (bidzyyys) [inline crates/motsu/src/lib.rs]:
```suggestion
        // Initial state: no events.
```
